### PR TITLE
.NET Isolated support + more efficient out-of-proc protocol

### PR DESCRIPTION
**Pull request by cgillum**
For .NET Isolated (and eventually Java), we're looking to use a more efficient out-of-process protocol compared to what we use for JS, Python, and PowerShell today. This PR contains support for the new protocol.

To summarize the changes in this PR:

* Adds a new Worker.Extensions.DurableTask project, which is the .NET Isolated extension for DF.
* Adds a new .NET 6 target to WebJobs.Extensions.DurableTask (required by the new OOProc gRPC dependencies)
* Add support for a local gRPC listener in the WebJobs extension (to be used instead of the local HTTP listener)
* Add "V2" middleware which implements the new OOProc protocol.
* Add an end-to-end test for .NET Isolated for use with GitHub Actions (currently "build only" due to a blocking bug).
* Added end-to-end tests for .NET in-process and Node.js to help ensure backwards compatibility

This change is intended to be 100% backwards compatible with existing OOProc support.

The 3 end-to-end tests introduced by these PRs run in GitHub Actions. I'm basically taking the smoke tests, putting them into a Docker container, running a basic "Hello Sequence" test, and using the HTTP status endpoint to verify that the orchestration completes successfully.

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation PR is ready to merge and referenced in `pending_docs.md`
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] I have added all required tests (Unit tests, E2E tests)
* [x] My changes **do not** require any extra work to be leveraged by OutOfProc SDKs
    * [ ] Otherwise: That work is being tracked here: #issue_or_pr_in_each_sdk

### CODE REVIEW

**Review by cgillum — COMMENTED** [review]
Adding comments that explain the changes in more detail.

File: src/WebJobs.Extensions.DurableTask/Bindings/BindingHelper.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/Bindings/BindingHelper.cs:43]: It would be good to write down this clarification on `RpcBaseUrl` as a comment. I can imagine it being helpful in an investigation.

File: src/WebJobs.Extensions.DurableTask/Bindings/OrchestrationTriggerAttributeBindingProvider.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/Bindings/OrchestrationTriggerAttributeBindingProvider.cs:90]: Does this have an effect on the expected contents of `function.json`? Namely, would it change the `binding->type` JSON field? Also, does this have implications for the development of the Java SDK?

File: src/WebJobs.Extensions.DurableTask/Bindings/OrchestrationTriggerAttributeBindingProvider.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/Bindings/OrchestrationTriggerAttributeBindingProvider.cs:148]: nit - the name `valueProvider` isn't terribly descriptive. Would `contextValueProvider` be more accurate?

File: src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/ContextImplementations/RemoteOrchestratorContext.cs:58]: Could this be a privacy concern? Unsure how likely it is for user-data to make it into the first 50 characters here

File: src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/DefaultPlatformInformation.cs:89]: Hmm, I don't love this, but I understand that C# doesn't otherwise have easy ways to map from string to `enum`.

File: src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs
Comment (davidmrdavid) [reply src/WebJobs.Extensions.DurableTask/LocalGrpcListener.cs:70]: I also learned this recently! From a David Fowler tweet :P

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:51]: Would we find ourselves skipping here if we're dealing with an entity? In any case, I think it would be great to document why it's safe to "skip" in here, instead of error'ing out

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:124]: Is there a need to check for the opening `{` character, or was this done just out of abundance of caution? If we're trying to make sure the return value is a JSON, I would think that simply trying to parse would be better check of schema-compliance, right?

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:241]: Same as above, how would we find ourselves in this method if not by invoking an activity?

File: src/Worker.Extensions.DurableTask/AssemblyInfo.cs
Comment (davidmrdavid) [reply src/Worker.Extensions.DurableTask/AssemblyInfo.cs:7]: @cgillum: can you add this to our extension release instructions?

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs
Comment (davidmrdavid) [inline src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:30]: What's the difference between these two styles of throwing?

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs
Comment (davidmrdavid) [inline src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:46]: Any reason why we don't already use `ASPNETCORE_FORWARDEDHEADERS_ENABLED`?

File: src/Worker.Extensions.DurableTask/DurableOrchestrator.cs
Comment (davidmrdavid) [inline src/Worker.Extensions.DurableTask/DurableOrchestrator.cs:83]: Is there a reason why we shouldn't incorporate this for-loop inside the constructor of `OrchestrationRuntimeState`?

File: test/Common/HttpApiHandlerTests.cs
Comment (davidmrdavid) [reply test/Common/HttpApiHandlerTests.cs:225]: I don't have context on this either, it would be good to investigate before merging.

File: test/SmokeTests/e2e-test.ps1
Comment (davidmrdavid) [inline test/SmokeTests/e2e-test.ps1:24]: Is there a reason why we don't build a Docker image with Azurite pre-installed in it? It might save time during automation...but I wouldn't block this PR based on that. Just an idea

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs
Comment (davidmrdavid) [inline src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:233]: Are we going to need this middleware for every API? If so, am I right in understanding that currently only `CallActivityAsync` is supported in dotnet-isolated`?

**Review by davidmrdavid — COMMENTED** [review]
Thanks for this PR. Since there's a fair bit of changes, I've left some preliminary thoughts and questions, and will be making a second pass after receiving your responses.

Thanks!

File: src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs
Comment (cgillum) [reply src/WebJobs.Extensions.DurableTask/OutOfProcMiddleware.cs:233]: I'm not sure what you mean by _every API_, but yes, we need this middleware to run for _every activity invocation_. I'm also not sure I understand your question about dotnet-isolated or how it relates to this middleware.

File: src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs
Comment (cgillum) [reply src/Worker.Extensions.DurableTask/DefaultDurableClientContext.cs:46]: I'm not sure. I think Connor was the one who looked into this originally. I'm just mentioning this here since it was mentioned in another GitHub issue. I don't currently know if it works or not.

File: test/SmokeTests/e2e-test.ps1
Comment (cgillum) [reply test/SmokeTests/e2e-test.ps1:24]: That would need to be a Docker image that's preconfigured for Azure Functions development scenarios. We wouldn't want this for production Docker images since that would bloat the image size and hurt cold start performance. This is probably a good question for the team that's working on the Azure Functions tooling experience, though (assuming we don't already have something like this).

**Review by cgillum — COMMENTED** [review]
Replied to PR comments.

**Review by davidmrdavid — APPROVED** [review]
LGTM **after** changing the "V2" suffix to something more descriptive and/or V3, as we already have a meaning for OOProc V2. Thanks!

File: .github/pull_request_template.md
Comment (cgillum) [reply .github/pull_request_template.md:20]: I think my original text is more accurate. It's not asking whether a code change they made warrants updating the version number. It's asking whether the PR author included a version change in the PR, in which case they must also update the Worker.Extensions.DurableTask's AssemblyInfo.cs file.

**Comment by cgillum** [comment]
Thanks @davidmrdavid. I went ahead and replaced `IsOutOfProcV2` with an enum, which hopefully makes things more descriptive.

File: .github/pull_request_template.md
Comment (davidmrdavid) [reply .github/pull_request_template.md:20]: Oh I completely misunderstood the original meaning. This makes sense, consider this issue resolved. Thanks!

**Comment by davidmrdavid** [comment]
@cgillum: I like the new design, it's good to merge for me!

**Comment by cgillum** [comment]
Thanks @davidmrdavid! I'll wait to merge until after the v2.6.1 release is complete.

**Pull request merged**
